core/src/main/java/org/apache/iceberg/io/datafile/DataFileServiceRegistry.java

```diff
@@ -0,0 +1,231 @@
+ */
+package org.apache.iceberg.io.datafile;
+
+import java.util.Map;
+import java.util.function.Function;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.avro.Avro;
+import org.apache.iceberg.common.DynMethods;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Objects;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Registry which provides the available {@link ReadBuilder}s and {@link WriteBuilder}s. Based on
+ * the `file format`, the required `data type` and the reader/writer `builderType` the registry
+ * returns the correct reader and writer builders. These builders could be used to generate the
+ * readers and writers.
+ *
+ * <p>File formats has to register the {@link ReadBuilder}s and the {@link AppenderBuilder}s which
+ * will be used to create the readers and the writers. The readers returned directly, the appenders
+ * are wrapped into a {@link WriteBuilder} and the {@link
+ * AppenderBuilder#build(AppenderBuilder.WriteMode)} method is used to finalize the appender
+ * configuration for the specific writer use-cases. The following inputs should be handled by the
+ * appender in the following cases:
+ *
+ * <ul>
+ *   <li>The appender's native input type - {@link AppenderBuilder.WriteMode#APPENDER}, {@link
+ *       AppenderBuilder.WriteMode#DATA_WRITER}, {@link
+ *       AppenderBuilder.WriteMode#EQUALITY_DELETE_WRITER}
+ *   <li>{@link org.apache.iceberg.deletes.PositionDelete} where the row is the appender's native
+ *       input type - {@link AppenderBuilder.WriteMode#POSITION_DELETE_WRITER}, {@link
+ *       AppenderBuilder.WriteMode#POSITION_DELETE_WITH_ROW_WRITER}
+ * </ul>
+ */
+public final class DataFileServiceRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(DataFileServiceRegistry.class);
+  // The list of classes which are used for registering the reader and writer builders
+  private static final String[] CLASSES_TO_REGISTER =
+      new String[] {
+        "org.apache.iceberg.parquet.Parquet",
+        "org.apache.iceberg.orc.ORC",
+        "org.apache.iceberg.arrow.vectorized.ArrowReader",
+        "org.apache.iceberg.flink.source.RowDataFileScanTaskReader",
+        "org.apache.iceberg.flink.sink.FlinkAppenderFactory",
+        "org.apache.iceberg.spark.source.DataFileServices"
+      };
+
+  private static final Map<Key, Function<EncryptedOutputFile, AppenderBuilder<?, ?>>>
+      APPENDER_BUILDERS = Maps.newConcurrentMap();
+  private static final Map<Key, Function<InputFile, ReadBuilder<?, ?>>> READ_BUILDERS =
+      Maps.newConcurrentMap();
+
+  /** Registers a new appender builder for the given format/input type. */
+  public static void registerAppender(
+      FileFormat format,
+      String inputType,
```

Comment:
`inputType` is a bit vague. How about `objectModelName` or something? It doesn't matter what this string is right?

Reply:
Done